pkg/tasks/repository_snapshot.go

```diff
@@ -284,6 +296,60 @@ func (sr *SnapshotRepository) lookupRepoObjects() (api.RepositoryResponse, error
 	return repoConfig, nil
 }
 
+func (sr *SnapshotRepository) cleanupOnCancel() error {
+	logger := LogForTask(sr.task.Id.String(), sr.task.Typename, sr.task.RequestID)
+	// TODO In Go 1.21 we could use context.WithoutCancel() to make copy of parent ctx that isn't canceled
+	ctxWithLogger := logger.WithContext(context.Background())
+	pulpClient := pulp_client.GetPulpClientWithDomain(ctxWithLogger, sr.domainName)
+	if sr.payload.SyncTaskHref != nil {
+		task, err := pulpClient.CancelTask(*sr.payload.SyncTaskHref)
+		if err != nil {
+			return err
```

Comment:
If you try to cancel a task that is already cancelled, you actually get a 409 Conflict response.  This causes it to bail out and not continue along.  We might wanna modify pulpClient.CancelTask to ignore those

Reply:
I think that's what I tried to do: https://github.com/content-services/content-sources-backend/blob/39801d186ecab35c300c08db729cd836d554498c/pkg/pulp_client/task.go#L48-L51

Are you seeing it return an error?

Reply:
I'll try to reproduce

Reply:
oh, i saw it in the logs, but and maybe i just assumed it errored.  Let me see if i can reproduce.

Reply:
Here's my attempt.  I created a repo with snapshotting =true, waited till it started publishing and then PATCH'd it to change the url.  

originally i was trying to cancel the publish, but it happens so fast, i'm usually too late.  But given that warning, i'd expect to see it twice
```
2:58PM DBG Stopped pulp task request_id=U7cBb4UJi3FTzrF7KHqJ5Bgx2CA54kzV state=completed task_href=/pulp/3a03f1bb/api/v3/tasks/018ae24d-462a-77e1-a408-34716ea65f90/ task_id=c6ba0b85-e54c-4bfc-a5da-6c1879b17c82 task_type=snapshot type=pulp_rpm.app.tasks.synchronizing.synchronize
2:58PM DBG Running pulp task request_id=U7cBb4UJi3FTzrF7KHqJ5Bgx2CA54kzV state=running task_href=/pulp/3a03f1bb/api/v3/tasks/018ae24e-96db-788c-87df-1505a9464055/ task_id=c6ba0b85-e54c-4bfc-a5da-6c1879b17c82 task_type=snapshot type=pulp_rpm.app.tasks.publishing.publish
2:58PM DBG Stopped pulp task request_id=U7cBb4UJi3FTzrF7KHqJ5Bgx2CA54kzV state=completed task_href=/pulp/3a03f1bb/api/v3/tasks/018ae24e-96db-788c-87df-1505a9464055/ task_id=c6ba0b85-e54c-4bfc-a5da-6c1879b17c82 task_type=snapshot type=pulp_rpm.app.tasks.publishing.publish
2:58PM DBG [Canceled Task] request_id=U7cBb4UJi3FTzrF7KHqJ5Bgx2CA54kzV task_id=c6ba0b85-e54c-4bfc-a5da-6c1879b17c82 task_type=snapshot
2:58PM INF [Enqueued Task] request_id=1upscvf4LU4wiq9HsC2TNqvZaib5B74l task_id=4321e385-85f4-470e-b1e5-1e10ca054767 task_type=snapshot
2:58PM INF [Dequeued Task] request_id=1upscvf4LU4wiq9HsC2TNqvZaib5B74l task_id=4321e385-85f4-470e-b1e5-1e10ca054767 task_type=snapshot
2:58PM INF [Enqueued Task] request_id=1upscvf4LU4wiq9HsC2TNqvZaib5B74l task_id=7174e380-8ea4-4a3c-8857-e90da01c3501 task_type=introspect
2:58PM INF ../../go/pkg/mod/github.com/ziflex/lecho/v3@v3.5.0/middleware.go:165 > bytes_in=77 bytes_out=591 host=localhost:8000 latency=38.371408 latency_human=38.371408ms method=PATCH referer= remote_ip=127.0.0.1 request_id=1upscvf4LU4wiq9HsC2TNqvZaib5B74l status=200 uri=/api/content-sources/v1.0/repositories/6be86308-fe56-42e7-a989-9f45864c1d6e user_agent="Apache-HttpClient/4.5.14 (Java/17.0.8)"
2:58PM INF [Dequeued Task] request_id=1upscvf4LU4wiq9HsC2TNqvZaib5B74l task_id=7174e380-8ea4-4a3c-8857-e90da01c3501 task_type=introspect
2:58PM INF Forcing introspection for 'https://jlsherrill.fedorapeople.org/fake-repos/needed-errata2/'
2:58PM DBG Introspecting https://jlsherrill.fedorapeople.org/fake-repos/needed-errata2/ request_id=1upscvf4LU4wiq9HsC2TNqvZaib5B74l task_id=7174e380-8ea4-4a3c-8857-e90da01c3501 task_type=introspect
2:58PM WRN Error 0 introspecting repository https://jlsherrill.fedorapeople.org/fake-repos/needed-errata2/ error="Error introspecting https://jlsherrill.fedorapeople.org/fake-repos/needed-errata2/: Cannot fetch https://jlsherrill.fedorapeople.org/fake-repos/needed-errata2/repodata/repomd.xml: 404" request_id=1upscvf4LU4wiq9HsC2TNqvZaib5B74l task_id=7174e380-8ea4-4a3c-8857-e90da01c3501 task_type=introspect
2:58PM DBG IntrospectionUrl returned 0 new packages request_id=1upscvf4LU4wiq9HsC2TNqvZaib5B74l task_id=7174e380-8ea4-4a3c-8857-e90da01c3501 task_type=introspect
2:58PM INF [Finished Task] task completed request_id=1upscvf4LU4wiq9HsC2TNqvZaib5B74l task_id=7174e380-8ea4-4a3c-8857-e90da01c3501 task_type=introspect
2:58PM ERR error cleaning up canceled snapshot error="409 Conflict" request_id=U7cBb4UJi3FTzrF7KHqJ5Bgx2CA54kzV task_id=c6ba0b85-e54c-4bfc-a5da-6c1879b17c82 task_type=snapshot
2:58PM INF [Finished Task] task canceled request_id=U7cBb4UJi3FTzrF7KHqJ5Bgx2CA54kzV task_id=c6ba0b85-e54c-4bfc-a5da-6c1879b17c82 task_type=snapshot
2:58PM DBG Running pulp task request_id=1upscvf4LU4wiq9HsC2TNqvZaib5B74l state=running task_href=/pulp/3a03f1bb/api/v3/tasks/018ae24e-a6f8-7cc4-a64d-fb2802fe32d1/ task_id=4321e385-85f4-470e-b1e5-1e10ca054767 task_type=snapshot type=pulp_rpm.app.tasks.synchronizing.synchronize
```

Actually i notice now that an error is logged, not a warn, so maybe its not hitting that conflict check?

Reply:
oh I see. Yeah I'll fix that.

Reply:
looks good!